plugins/httpapi/aci.py

```diff
@@ -0,0 +1,153 @@
+    a connection to the ACI controller, send API requests and process the
+    response from the controller.
+version_added: "2.1.0"
+"""
+
+
+import json
+
+from ansible.module_utils._text import to_text
+from ansible.module_utils.connection import ConnectionError
+from ansible.plugins.httpapi import HttpApiBase
+
+
+class HttpApi(HttpApiBase):
+
+    def __init__(self, *args, **kwargs):
+        super(HttpApi, self).__init__(*args, **kwargs)
+        self.headers = {
+            'Content-Type': "application/json"
+        }
+        self.auth = None
+        self.aci_host = None
+        self.aci_user = None
+        self.aci_pass = None
+
+    def get_auth(self, auth, host, username, password):
+        self.auth = auth
+        self.aci_host = host
+        self.aci_user = username
+        self.aci_pass = password
+
+    def login(self, username, password):
+        ''' Log in to APIC '''
+
+        # Perform login request
+        method = 'POST'
+        path = '/api/aaaLogin.json'
+        payload = {'aaaUser': {'attributes': {'name': username, 'pwd': password}}}
+        data = json.dumps(payload)
+        try:
+            response, response_data = self.connection.send(path, data, method=method, headers=self.headers, timeout=10)
+            response_value = self._get_response_value(response_data)
+            self.connection._auth = {'Cookie': 'APIC-Cookie={0}'
+                                     .format(self._response_to_json(response_value).get('imdata')[0]['aaaLogin']['attributes']['token'])}
+
+        except Exception as e:
+            msg = 'Error on attempt to connect and authenticate with user: {0} to APIC: {1}. {2} '.format(username, self.connection.get_option("host"), e)
+            raise ConnectionError(msg)
+
+    def logout(self):
+        method = 'POST'
+        path = '/api/aaaLogout.json'
+
+        try:
+            response, response_data = self.connection.send(path, {}, method=method, headers=self.headers, force_basic_auth=True)
+        except Exception as e:
+            msg = 'Error on attempt to logout from APIC. {0}'.format(e)
+            raise ConnectionError(self._return_info(None, method, path, msg))
+
+        self._verify_response(response, method, path, response_data)
+        # Clean up tokens
+        self.connection._auth = None
+
+    def send_request(self, method, path, json=None):
+        ''' This method handles all APIC REST API requests other then login '''
+        if json is None:
+            json = {}
+
+        if self.auth is not None:
+            self.connection._auth = {'Cookie': '{0}'
+                                     .format(self.auth)}
+
+        if self.aci_host is not None:
+            self.connection.set_option("host", self.aci_host)
+
+        if self.aci_user is not None:
+            self.connection.set_option("remote_user", self.aci_user)
+
+        if self.aci_pass is not None:
+            self.connection.set_option("password", self.aci_pass)
+
```

Comment:
How do we deal with the other settings like proxy, ssl, port, ...?

Reply:
I'll include them using set option before next push